src/analysis/normed_space/normed_group_hom.lean

```diff
@@ -0,0 +1,358 @@
+
+@[simp] lemma coe_mk (f) (h₁) (h₂) (h₃) : ⇑(⟨f, h₁, h₂, h₃⟩ : normed_group_hom V₁ V₂) = f := rfl
+
+@[simp] lemma coe_to_add_monoid_hom : ⇑f.to_add_monoid_hom = f := rfl
+
+lemma to_add_monoid_hom_injective :
+  function.injective (@normed_group_hom.to_add_monoid_hom V₁ V₂ _ _) :=
+λ f g h, ext $ show ⇑f.to_add_monoid_hom = g, by { rw h, refl }
+
+@[simp] lemma mk_to_add_monoid_hom (f) (h₁) (h₂) (h₃) :
+  (⟨f, h₁, h₂, h₃⟩ : normed_group_hom V₁ V₂).to_add_monoid_hom = ⟨f, h₁, h₂⟩ := rfl
+
+@[simp] lemma map_zero : f 0 = 0 := f.to_add_monoid_hom.map_zero
+
+@[simp] lemma map_add (x y) : f (x + y) = f x + f y := f.to_add_monoid_hom.map_add _ _
+
+@[simp] lemma map_sum {ι : Type*} (v : ι → V₁) (s : finset ι) :
+  f (∑ i in s, v i) = ∑ i in s, f (v i) :=
+f.to_add_monoid_hom.map_sum _ _
+
+@[simp] lemma map_sub (x y) : f (x - y) = f x - f y := f.to_add_monoid_hom.map_sub _ _
+
+@[simp] lemma map_neg (x) : f (-x) = -(f x) := f.to_add_monoid_hom.map_neg _
+
+/-- Make a normed group hom from a group hom and a norm bound. -/
+def mk' (f : V₁ →+ V₂) (C : ℝ≥0) (hC : ∀ x, ∥f x∥ ≤ C * ∥x∥) : normed_group_hom V₁ V₂ :=
+{ bound' := ⟨C, hC⟩ .. f}
+
+@[simp] lemma coe_mk' (f : V₁ →+ V₂) (C) (hC) : ⇑(mk' f C hC) = f := rfl
+
+/-- Predicate asserting a norm bound on a normed group hom. -/
+def bound_by (f : normed_group_hom V₁ V₂) (C : ℝ≥0) : Prop := ∀ x, ∥f x∥ ≤ C * ∥x∥
+
+lemma mk'_bound_by (f : V₁ →+ V₂) (C) (hC) : (mk' f C hC).bound_by C := hC
+
+lemma bound : ∃ C, 0 < C ∧ f.bound_by C :=
+begin
+  obtain ⟨C, hC⟩ := f.bound',
+  let C' : ℝ≥0 := ⟨max C 1, le_max_right_of_le zero_le_one⟩,
+  use C',
+  simp only [C', ← nnreal.coe_lt_coe, subtype.coe_mk, nnreal.coe_zero,
+    lt_max_iff, zero_lt_one, or_true, true_and],
+  intro v,
+  calc ∥f v∥
+      ≤ C * ∥v∥ : hC v
+  ... ≤ max C 1 * ∥v∥ : mul_le_mul (le_max_left _ _) le_rfl (norm_nonneg _) _,
+  exact zero_le_one.trans (le_max_right _ _)
+end
+
+lemma lipschitz_of_bound_by (C : ℝ≥0) (h : f.bound_by C) :
+  lipschitz_with (nnreal.of_real C) f :=
+lipschitz_with.of_dist_le' $ λ x y, by simpa only [dist_eq_norm, f.map_sub] using h (x - y)
+
+theorem antilipschitz_of_bound_by {K : ℝ≥0} (h : ∀ x, ∥x∥ ≤ K * ∥f x∥) :
+  antilipschitz_with K f :=
+antilipschitz_with.of_le_mul_dist $
+λ x y, by simpa only [dist_eq_norm, f.map_sub] using h (x - y)
+
+protected lemma uniform_continuous (f : normed_group_hom V₁ V₂) :
+  uniform_continuous f :=
+begin
+  obtain ⟨C, C_pos, hC⟩ := f.bound,
+  exact (lipschitz_of_bound_by f C hC).uniform_continuous
+end
+
+@[continuity]
+protected lemma continuous (f : normed_group_hom V₁ V₂) : continuous f :=
+f.uniform_continuous.continuous
+
+variables {f g}
+
+instance : has_zero (normed_group_hom V₁ V₂) :=
+⟨{ bound' := ⟨0, λ v, show ∥(0 : V₂)∥ ≤ 0 * _, by rw [norm_zero, zero_mul]⟩,
+   .. (0 : V₁ →+ V₂) }⟩
+
+instance : inhabited (normed_group_hom V₁ V₂) := ⟨0⟩
+
+lemma coe_inj ⦃f g : normed_group_hom V₁ V₂⦄ (h : (f : V₁ → V₂) = g) : f = g :=
```

Comment:
Don't you already have a `coe_inj` above?

Reply:
This one is now gone